src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -44,69 +50,105 @@ class IcebergParquetValueParser {
+    return parseColumnValueToParquet(
+        value, type, statsMap, defaultTimezone, insertRowsCurrIndex, null, false);
+  }
+
+  private static ParquetBufferValue parseColumnValueToParquet(
+      Object value,
+      Type type,
+      Map<String, RowBufferStats> statsMap,
+      ZoneId defaultTimezone,
+      long insertRowsCurrIndex,
+      String path,
+      boolean isdDescendantsOfRepeatingGroup) {
+    path = (path == null || path.isEmpty()) ? type.getName() : path + "." + type.getName();
     float estimatedParquetSize = 0F;
     if (value != null) {
-      estimatedParquetSize += ParquetBufferValue.DEFINITION_LEVEL_ENCODING_BYTE_LEN;
-      PrimitiveType primitiveType = type.asPrimitiveType();
-      switch (primitiveType.getPrimitiveTypeName()) {
-        case BOOLEAN:
-          int intValue =
-              DataValidationUtil.validateAndParseBoolean(
-                  type.getName(), value, insertRowsCurrIndex);
-          value = intValue > 0;
-          stats.addIntValue(BigInteger.valueOf(intValue));
-          estimatedParquetSize += ParquetBufferValue.BIT_ENCODING_BYTE_LEN;
-          break;
-        case INT32:
-          int intVal = getInt32Value(value, primitiveType, insertRowsCurrIndex);
-          value = intVal;
-          stats.addIntValue(BigInteger.valueOf(intVal));
-          estimatedParquetSize += 4;
-          break;
-        case INT64:
-          long longVal = getInt64Value(value, primitiveType, defaultTimezone, insertRowsCurrIndex);
-          value = longVal;
-          stats.addIntValue(BigInteger.valueOf(longVal));
-          estimatedParquetSize += 8;
-          break;
-        case FLOAT:
-          float floatVal =
-              (float)
-                  DataValidationUtil.validateAndParseReal(
-                      type.getName(), value, insertRowsCurrIndex);
-          value = floatVal;
-          stats.addRealValue((double) floatVal);
-          estimatedParquetSize += 4;
-          break;
-        case DOUBLE:
-          double doubleVal =
-              DataValidationUtil.validateAndParseReal(type.getName(), value, insertRowsCurrIndex);
-          value = doubleVal;
-          stats.addRealValue(doubleVal);
-          estimatedParquetSize += 8;
-          break;
-        case BINARY:
-          byte[] byteVal = getBinaryValue(value, primitiveType, stats, insertRowsCurrIndex);
-          value = byteVal;
-          estimatedParquetSize +=
-              ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN + byteVal.length;
-          break;
-        case FIXED_LEN_BYTE_ARRAY:
-          byte[] fixedLenByteArrayVal =
-              getFixedLenByteArrayValue(value, primitiveType, stats, insertRowsCurrIndex);
-          value = fixedLenByteArrayVal;
-          estimatedParquetSize +=
-              ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN + fixedLenByteArrayVal.length;
-          break;
-        default:
+      if (type.isPrimitive()) {
+        if (!statsMap.containsKey(path)) {
```

Comment:
collapse the two map hits (across containsKey and get) into one, like so:
```
RowBufferStats stats = statsMap.get();
if (stats == null) {
    throw
}
```

Reply:
Moved the `containsKey` check to the beginning of the method.